src/frontend/src/optimizer/plan_node/logical_multi_join.rs

```diff
@@ -483,6 +484,208 @@ impl LogicalMultiJoin {
         Ok(join_ordering)
     }
 
+    pub fn as_bushy_tree_join(&self) -> Result<PlanRef> {
+        // Join tree internal representation
+        #[derive(Clone, Default, Debug)]
+        struct JoinTreeNode {
+            idx: Option<usize>,
+            left: Option<Box<JoinTreeNode>>,
+            right: Option<Box<JoinTreeNode>>,
+            height: usize,
+        }
+
+        // join graph internal representation
+        #[derive(Clone, Debug)]
+        struct GraphNode {
+            id: usize,
+            join_tree: JoinTreeNode,
+            relations: HashSet<usize>,
+        }
+
+        let mut nodes: HashMap<_, _> = (0..self.inputs.len())
+            .map(|idx| GraphNode {
+                id: idx,
+                relations: HashSet::new(),
+                join_tree: JoinTreeNode {
+                    idx: Some(idx),
+                    left: None,
+                    right: None,
+                    height: 0,
+                },
+            })
+            .enumerate()
+            .collect();
+        let (eq_join_conditions, _) = self
+            .on
+            .clone()
+            .split_by_input_col_nums(&self.input_col_nums(), true);
+
+        for ((src, dst), _) in eq_join_conditions {
+            nodes.get_mut(&src).unwrap().relations.insert(dst);
+            nodes.get_mut(&dst).unwrap().relations.insert(src);
+        }
+
+        let mut optimized_bushy_tree = None;
+        let mut que = VecDeque::from([nodes]);
+        let mut isolated = HashSet::new();
+
+        while let Some(mut nodes) = que.pop_front() {
+            if nodes.len() == 1 {
+                let node = nodes.into_values().next().unwrap();
+                optimized_bushy_tree = Some(optimized_bushy_tree.map_or(
+                    node.clone(),
+                    |old_tree: GraphNode| {
+                        if node.join_tree.height < old_tree.join_tree.height {
+                            node
+                        } else {
+                            old_tree
+                        }
+                    },
+                ));
+                continue;
+            }
+
+            let (idx, _) = nodes.iter().min_by_key(|(_, g)| g.relations.len()).unwrap();
+
+            let n = nodes.remove(&idx.clone()).unwrap();
+
+            if n.relations.is_empty() {
+                isolated.insert(n.id);
+                que.push_back(nodes);
+                continue;
+            }
+
+            for merge_node in &n.relations {
```

Comment:
The crucial factor in selecting the `merge_node` lies in its priority. In my opinion, it would be wise to merge join trees with lower height.